### PR TITLE
Guess firmware format function and docs

### LINKED ISSUE #1174: Add guessFirmwareFormat function
Actually guessFirmwareFormat is not present in zwave-js

### PR DESCRIPTION
**Pull request by robertsLando**
Fixes #1174

### CODE REVIEW

File: docs/api/node.md [hunk 1]
```diff
@@ -132,7 +132,17 @@ Starts an OTA firmware update process for this node. This method takes two argum
 -   `data` - A buffer containing the firmware image in a format supported by the device
 -   `target` - _(optional)_ The firmware target (i.e. chip) to upgrade. 0 updates the Z-Wave chip, >=1 updates others if they exist
 
-The library includes a helper method (exported from `zwave-js/Utils`) to extract the raw firmware image from some common file types:
+The library includes 2 helper method (exported from `zwave-js/Utils`) to guess the firmware format and extract the raw firmware image from some common file types:
+
+```ts
+function guessFirmwareFormat(
+	filename: string,
+	firmware: Buffer,
```
Comment (AlCalzone) [inline docs/api/node.md]: I guess we should call this parameter `rawData` to match the example below

File: docs/api/node.md [hunk 2]
```diff
@@ -155,6 +165,38 @@ interface Firmware {
 
 If no firmware data can be extracted, the method will throw.
 
+Example usage:
+
+```ts
+// Extract the firmware from the uploaded file
+const rawData = Buffer.from(firmware);
```
Comment (AlCalzone) [inline docs/api/node.md]: Hmm I think this is confusing if you read this the first time. Maybe just state that `rawData` is a Buffer with the file contents and `filename` the corresponding filename?

File: docs/api/node.md [hunk 3]
```diff
@@ -132,7 +132,17 @@ Starts an OTA firmware update process for this node. This method takes two argum
 -   `data` - A buffer containing the firmware image in a format supported by the device
 -   `target` - _(optional)_ The firmware target (i.e. chip) to upgrade. 0 updates the Z-Wave chip, >=1 updates others if they exist
 
-The library includes a helper method (exported from `zwave-js/Utils`) to extract the raw firmware image from some common file types:
+The library includes 2 helper method (exported from `zwave-js/Utils`) to guess the firmware format and extract the raw firmware image from some common file types:
```
Comment (AlCalzone) [inline docs/api/node.md]: I know I originally wrote this, but we should mention which input files we support.

File: packages/core/src/util/firmware.ts [hunk 4]
```diff
@@ -9,6 +9,28 @@ export interface Firmware {
...
+	firmware: Buffer,
+): FirmwareFileFormat {
+	if (
+		(filename.endsWith(".exe") || filename.endsWith(".ex_")) &&
+		firmware.includes(Buffer.from("Aeon Labs", "utf8"))
+	) {
+		return "aeotec";
+	} else if (/\.(hex|ota|otz)$/.test(filename)) {
+		return filename.slice(-3) as FirmwareFileFormat;
+	} else {
```
Comment (AlCalzone) [inline packages/core/src/util/firmware.ts:34]: Should we throw a separate error to let users know that `.hec` is not supported?

File: packages/core/src/util/firmware.ts [hunk 5]
```diff
@@ -9,6 +9,28 @@ export interface Firmware {
...
+		throw new Error("could not detect firmware format");
```
Comment (AlCalzone) [inline packages/core/src/util/firmware.ts]: This should be a `ZWaveError` with a corresponding `ZWaveErrorCode`.

File: docs/api/node.md [hunk 6]
```diff
@@ -132,7 +132,17 @@ Starts an OTA firmware update process for this node. This method takes two argum
...
+
+```ts
+function guessFirmwareFormat(
```
Comment (AlCalzone) [inline docs/api/node.md]: nit: We should probably call this `guessFirmwareFileFormat` (non-breaking, preferred) or rename `FimwareFileFormat` to `FirmwareFormat` (breaking).

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: docs/api/node.md (hunk 2 above)
Comment (robertsLando) [reply docs/api/node.md]: Yep done

File: docs/api/node.md (hunk 3 above)
Comment (robertsLando) [reply docs/api/node.md]: It's mentioned later on extractFirmware function

File: packages/core/src/util/firmware.ts (hunk 4 above)
Comment (robertsLando) [reply packages/core/src/util/firmware.ts:34]: Humm nope, it's invalid like the others for now until we don't support it

File: packages/core/src/util/firmware.ts (hunk 5 above)
Comment (robertsLando) [reply packages/core/src/util/firmware.ts]: fixed

File: docs/api/node.md (hunk 6 above)
Comment (robertsLando) [reply docs/api/node.md]: Done

File: packages/core/src/util/firmware.ts (hunk 4 above)
Comment (AlCalzone) [reply packages/core/src/util/firmware.ts:34]: I mean saying `.hec` is not (and probably won't be supported) is clearer than "we don't know what this is, maybe we could support it".

**Review by AlCalzone — APPROVED** [review]

**Pull request merged**